File: src/my_cpp_pkg/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"

class number_publisher : public rclcpp::Node
{
    public:
        number_publisher() : Node("number_publisher")
        {
            RCLCPP_INFO(this->get_logger(), "Publisher node is initialised");
            publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
            timer_ = this->create_wall_timer(std::chrono::milliseconds(500), std::bind(&number_publisher::publish_number, this));
            
        }

    private:

        void publish_number(){
           auto msg = example_interfaces::msg::Int64();
           counter_ += 1;
           msg.data = counter_;
           publisher_->publish(msg);
        }

        rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
        rclcpp::TimerBase::SharedPtr timer_;
        int counter_ = 0;
};

int main(int argc, char **argv)
{

    rclcpp::init(argc, argv);
    auto node = std::make_shared<number_publisher>();

    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64

class number_publisher(Node):
    def __init__(self):
        super().__init__("number_publisher")

        self.publisher_ = self.create_publisher(Int64, "number", 10)
        self.msg_ = Int64()
        self.counter_ = 0
        self.create_timer(1/3, self.callback_counterloop)
        self.get_logger().info("number publisher node initialised")
        
    def callback_counterloop(self):
        self.counter_ += 1
        self.msg_.data = self.counter_
        self.publisher_.publish(self.msg_)

def main(args=None):
    rclpy.init(args=args)

    node = number_publisher()

    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()